src/bin/migrate.rs

```diff
@@ -0,0 +1,127 @@
+//! This binary is used to migrate user IDs from base64 to hex encoding.
+use std::{path::Path, str::FromStr};
+
+use anyhow::{anyhow, Context, Result};
+use base64::{engine::general_purpose, Engine as _};
+use famedly_sync::Config;
+use futures::StreamExt;
+use tracing::level_filters::LevelFilter;
+use zitadel_rust_client::v2::{
+	users::{
+		ListUsersRequest, SearchQuery, SetHumanProfile, TypeQuery, UpdateHumanUserRequest,
+		UserFieldName, Userv2Type,
+	},
+	Zitadel,
+};
+
+#[tokio::main]
+async fn main() -> Result<()> {
+	// Config
+	let config_path =
+		std::env::var("FAMEDLY_SYNC_CONFIG").unwrap_or_else(|_| "./config.yaml".to_owned());
+	let config = Config::new(Path::new(&config_path))?;
+
+	// Tracing
+	let subscriber = tracing_subscriber::FmtSubscriber::builder()
+		.with_max_level(
+			config
+				.log_level
+				.as_ref()
+				.map_or(Ok(LevelFilter::INFO), |s| LevelFilter::from_str(s))?,
+		)
+		.finish();
+	tracing::subscriber::set_global_default(subscriber)
+		.context("Setting default tracing subscriber failed")?;
+
+	tracing::info!("Starting migration");
+	tracing::debug!("Old external IDs will be base64 decoded and re-encoded as hex");
+	tracing::debug!("Note: External IDs are stored in the nick_name field of the user's profile in Zitadel, often referred to as uid.");
+
+	// Zitadel
+	let zitadel_config = config.zitadel.clone();
+	let mut zitadel = Zitadel::new(zitadel_config.url, zitadel_config.key_file).await?;
+
+	// Get all users
+	let mut stream = zitadel.list_users(
+		ListUsersRequest::new(vec![
+			SearchQuery::new().with_type_query(TypeQuery::new(Userv2Type::Human))
+		])
+		.with_asc(true)
+		.with_sorting_column(UserFieldName::NickName),
+	)?;
+
+	// Process each user
+	while let Some(user) = stream.next().await {
+		let user_id = user.user_id().ok_or_else(|| anyhow!("User lacks ID"))?;
+		let human = user.human().ok_or_else(|| anyhow!("User not human"))?;
+		let profile = human.profile().ok_or_else(|| anyhow!("User lacks profile"))?;
+
+		let given_name = profile.given_name();
+		let family_name = profile.family_name();
+		let display_name = profile.display_name();
+
+		let Some(nick_name) = profile.nick_name() else {
+			tracing::warn!(user_id = %user_id, "User lacks nick_name (external uid), skipping");
+			continue;
+		};
+
+		tracing::info!(user_id = %user_id, old_uid = %nick_name, "Starting migration for user");
+
+		let new_external_id = if nick_name.is_empty() {
+			// Keep empty string as is, will skip it later
+			nick_name.to_string()
+		} else {
+			// First check if it's valid hex
+			if nick_name.chars().all(|c| c.is_ascii_hexdigit()) && nick_name.len() % 2 == 0 {
+				// If valid hex, keep as is
+				tracing::debug!(  user_id = %user_id, old_uid = %nick_name,"Valid hex uid detected, keeping as is");
```

Comment:
The valid set of characters for a hex-encoded string is a subset of the ones for a base64 encoded string. As such, while unlikely, this could end up mis-identifying some base64-encoded IDs as hex encoded.

We can still do this kind of check, but we should probably employ it as a heuristic (that can be overridden) on the whole database instead of per-record.